File: Julia/challenge_0.jl
using Sockets

function handle_client(socket::TCPSocket)
    println("New client connected: $(getsockname(socket)[1])")
    try
        while isopen(socket)
            data = readavailable(socket)
            if isempty(data)
                println("Client ended: $(getsockname(socket)[1])")
                break
            end
            write(socket, data)
        end
    catch e
        println("Client error: $e")
    finally
        close(socket)
        println("Client disconnected: $(getsockname(socket)[1])")
    end
end

function start_server(port::Int=40000)
    server = listen(IPv4(0), port)
    println("Server listening at 0.0.0.0:$port")
    
    while true
        socket = accept(server)
        @async handle_client(socket)
    end
end

# Start the server
start_server(40000)

File: Perl/challenge_0.pl
#!/usr/bin/perl
use strict;
use warnings;
use IO::Socket::INET;

# Configuration
my $port = 40000;
my $buffer_size = 1024;

# Create TCP server socket
my $server = IO::Socket::INET->new(
    LocalPort => $port,
    Proto     => 'tcp',
    Listen    => 10,
    ReuseAddr => 1,
) or die "Cannot create server socket: $!\n";

print "Server listening on port $port\n";

# Main server loop
while (my $client = $server->accept()) {
    my $peer_addr = $client->peerhost();
    my $peer_port = $client->peerport();
    print "[@{[scalar localtime]}] Connected: $peer_addr:$peer_port\n";

    # Fork a new process to handle the client
    if (fork() == 0) {
        # Child process
        while (1) {
            my $data;
            my $bytes_read = $client->recv($data, $buffer_size);
            if (!defined $bytes_read || length($data) == 0) {
                # Client disconnected
                print "[@{[scalar localtime]}] Disconnected: $peer_addr:$peer_port\n";
                last;
            }
            # Echo data back
            $client->send($data) or do {
                print "[@{[scalar localtime]}] Send error: $!\n";
                last;
            };
        }
        $client->close();
        exit; # Exit child process
    }
    $client->close(); # Parent closes client socket
}

# Cleanup
$server->close();
